test: add reverts test documentation (#1143)
ApeWorX/ape: docs/userguides/testing.md

    assert actual == expect
```

## Testing Transaction Failures

Similar to `pytest.raises()`, you can use `ape.reverts()` to assert that contract transactions fail and revert.

From our earlier example we can see this in action:

```python
def test_authorization(my_contract, owner, not_owner):
    my_contract.set_owner(sender=owner)
    assert owner == my_contract.owner()

    with ape.reverts("!authorized"):
        my_contract.authorized_method(sender=not_owner)
```

`reverts()` takes two optional parameters:

### `expected_message`

This is the expected revert reason given when the transaction fails.
If the message in the `ContractLogicError` raised by the transaction failure is empty or does not match the `expected_message`, then `ape.reverts()` will raise an `AssertionError`.

### `dev_message`

This is the expected dev message corresponding to the line in the contract's source code where the error occurred.

> Dev messages take the form of a comment in Vyper, and should be placed on the line that may cause a transaction revert:

Reviewer: Should we include a note about how this is used for gas optimization and keeping revert reasons shorter?

Author: Updated in 71d5cc10

```python
assert x != 0 # dev: invalid value
```

Take for example:

```python
# @version 0.3.7

@external
def check_value(_value: uint256) -> bool:

>     assert _value != 0 # dev: invalid value

Reviewer: To match default python lint rules, should we put 2 spaces between comment and code?

Does this matter in terms of detecting dev reverts?

Author: Updated in 101f372a

> Does this matter in terms of detecting dev reverts?

No, we basically just check for `#\s*(dev:.*)` regardless of what precedes it

    return True
```

> We can explicitly cause a transaction revert and check that it was that assertion line that failed by supplying an expected `dev_message`:

Reviewer 2: ```suggestion
We can explicitly cause a transaction revert and check the failed line by supplying an expected `dev_message`:
```

```python
def test_authorization(my_contract, owner):
    with ape.reverts(dev_message="dev: invalid value"):
        my_contract.check_value(sender=owner)
```

> When the transaction reverts and `ContractLogicError` is raised, `ape.reverts()` will check the source contract and see if the line that failed contains the message.

Reviewer 2: ```suggestion
When the transaction reverts and `ContractLogicError` is raised, `ape.reverts()` will check the source contract to see if the failed line contains a message.
```

There are a few scenarios where `AssertionError` will be raised when using `dev_message`:
- If the line in the source contract has a different dev message or no dev message
- If the contract source cannot be obtained
- If the transaction trace cannot be obtained

> Because `dev_message` relies on transaction tracing to function, you must use a provider like `ape-hardhat` when testing with `dev_message`.

Reviewer: We can link to ape-hardhat here

Author: Updated in 68372465 including links for `ape-vyper`, `ape-solidity`, and `ape-cairo` in my other paragraph

### Caveats

#### Language Support

As of `ape` version `0.5.6`, `dev_messages` assertions are only available with Vyper and `ape-vyper`.

> You must use `ape-vyper` to compile your contracts for the relevant source information to be present for `ape.reverts()`.

Reviewer: I think we should rephrase this a bit. Shortly after we release, some dedicated dev could add dev-str detection to their super secret language plugin, in which case this statement becomes false.

Author: Updated in 28015695

I basically rephrased it to say "as of 0.5.6 this works with ape-vyper, but not ape-solidity or ape-cairo"

That way it's specific to which of our plugins don't have support yet, rather than saying only `ape-vyper` can ever be used
